servers/dsub/jobs/controllers/utils/generator.py

```diff
@@ -0,0 +1,164 @@
+            create_time_max (datetime): the latest job create time
+            offset_id (): paginator offset id
+
+        Returns:
+            generator: Retrieved jobs
+    """
+    proj_id = query.extensions.project_id if query.extensions else None
+    dstat_params = query_parameters.api_to_dsub(query)
+
+    # If create_time_max is not set, but we have to client-side filter by
+    # end-time, set create_time_max = query.end because all the jobs with
+    # create_time >= query.end cannot possibly match the query.
+    if not create_time_max and query.end:
+        create_time_max = query.end
+
+    # If submission time is not specified, set the create_time_min to (start time - 7days)
+    # to avoid iterating through the whole job list.
+    create_time_min = dstat_params.get('create_time')
+    if not create_time_min and query.start:
+        create_time_min = query.start - datetime.timedelta(
+            days=_MAX_PENDING_TIME_DAYS)
+
+    jobs = execute_redirect_stdout(lambda: dstat.lookup_job_tasks(
+        provider=provider,
+        statuses=dstat_params['statuses'],
+        user_ids=dstat_params.get('user_ids'),
+        job_ids=dstat_params.get('job_ids'),
+        task_ids=dstat_params.get('task_ids'),
+        create_time_min=create_time_min,
+        create_time_max=create_time_max,
+        job_names=dstat_params.get('job_names'),
+        labels=dstat_params.get('labels')))
+
+    last_create_time = None
+    job_buffer = []
+    for j in jobs:
+        job = _query_jobs_result(j, proj_id)
+        # Filter pending vs. running jobs since dstat does not have
+        # a corresponding status (both RUNNING)
+        if query.statuses and job.status not in query.statuses:
+            continue
+        if query.start and (not job.start or job.start < query.start):
+            continue
+        if query.end and (not job.end or job.end > query.end):
+            continue
+
+        # If this job is from the last page, skip it and continue generating
+        if create_time_max and job.submission == create_time_max:
+            if offset_id and job.id < offset_id:
+                continue
+
+        # Build up a buffer of jobs with the same create time. Once we get a
+        # job with an older create time we yield all the jobs in the buffer
+        # sorted by job-id + task-id
+        job_buffer.append(job)
+        if job.submission != last_create_time:
+            for j in sorted(job_buffer, key=lambda j: j.id):
+                yield j
+            job_buffer = []
+        last_create_time = job.submission
+
+    # If we hit the end of the dstat job generator, ensure to yield the jobs
+    # stored in the buffer before returning
+    for j in sorted(job_buffer, key=lambda j: j.id):
+        yield j
+
+
+def generate_jobs_count(provider,
+                        project_id,
+                        window_min,
+                        window_max=datetime.datetime.now(tz=tzlocal())):
```

Comment:
I think it would be better to default this to `None` in general. Also complex default parameters like this can have surprising behaviors; I would need to experiment and reread the Python docs to check that this does the right thing.

Reply:
Changed to None.